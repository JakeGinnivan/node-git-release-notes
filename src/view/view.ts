import * as fs from 'fs-extra'
import glob from '../utils/glob'
import { fromMarkdown } from '../release-notes/read'
import { formatVersionChanges, defaultFormattingInfo } from '../release-notes/write'
import { ChangeLogItem } from '../release-notes/model'

export const view = async (fileGlob: string, versionSpec: string) => {
    const files = await glob(fileGlob, {
        nocase: true,
        ignore: 'node_modules/**',
    })

    if (files.length !== 1) {
        // tslint:disable-next-line:no-string-throw
        throw `Expected to find one file matching ${fileGlob}`
    }

    const changelogContents = (await fs.readFile(files[0])).toString()

    return extractChanges(files[0], changelogContents, versionSpec)
}

export const extractChanges = (filename: string, changelog: string, versionSpec: string) => {
    const releaseNotes = fromMarkdown(changelog, filename)

    const parts = versionSpec.split('...')
    if (parts.length === 1) {
        const version = releaseNotes.versions.find(v =>
            v.version.toLowerCase().includes(parts[0].toLowerCase()))
        if (!version) {
            // tslint:disable-next-line:no-string-throw
            throw `Cannot find version information for ${parts[0]}`
        }

        return formatVersionChanges(
            version.summary,
            version.changeLogs,
            releaseNotes.formattingData || defaultFormattingInfo,
        )
    }

    if (parts.length === 2) {
        let fromIsExclusive = false
        // We have a range, is it open ended?
        const range = {
            from: parts[0].toLowerCase(),
            to: parts[1].toLowerCase(),
        }

        if (range.from[0] === '[' && range.from[range.from.length - 1] === ']') {
            fromIsExclusive = true
            range.from = range.from.slice(1, range.from.length - 1)
        }

        let toFound = false
        let fromFound = false

        type Acc = {
            changes: ChangeLogItem[],
            summaries: string[],
        }
        const allChanges = releaseNotes.versions.reduceRight<Acc>((acc, val) => {
            if (range.to && val.version.toLowerCase().includes(range.to)) {
                toFound = true
            }
            if (!range.from || val.version.toLowerCase().includes(range.from)) {
                fromFound = true
                if (fromIsExclusive) {
                    return acc
                }
            }

            if (fromFound && !toFound) {
                acc.changes.push(...val.changeLogs)
                if (val.summary) { acc.summaries.push(val.summary) }
            }

            return acc
        }, { changes: [], summaries: [] })

        return formatVersionChanges(
            allChanges.summaries.join('\n'),
            allChanges.changes,
            releaseNotes.formattingData || defaultFormattingInfo,
        )
    }

    // tslint:disable-next-line:no-string-throw
    throw 'Spec is not valid, should be in format <version> or [<fromversion>]...[<toversion>]'
}
