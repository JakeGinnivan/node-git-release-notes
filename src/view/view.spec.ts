import { extractChanges } from './view'

it('Can get changes for a single version', () => {
    const changes = extractChanges('file.md', `# Changelog

## v2.0.0
- Change 1
- Change 2
  - Change 2.1

## v1.0.0
- Change 0
`, '2.0.0')

    expect(changes).toMatchSnapshot()
})

it('Can get changes after a version', () => {
    const changes = extractChanges('file.md', `# Changelog

## v3.0.0
- Change 3
- Change 4

## v2.0.0
- Change 1
- Change 2
  - Change 2.1

## v1.0.0
- Change 0
`, '2.0.0...')

    expect(changes).toMatchSnapshot()
})

it('Can get changes before a version', () => {
    const changes = extractChanges('file.md', `# Changelog

## v3.0.0
- Change 3
- Change 4

## v2.0.0
- Change 1
- Change 2
  - Change 2.1

## v1.0.0
- Change 0
`, '...3.0.0')

    expect(changes).toMatchSnapshot()
})

it('Can get changes between versions', () => {
    const changes = extractChanges('file.md', `# Changelog

## v3.0.0
- Change 3
- Change 4

## v2.0.1
- Fix 1

## v2.0.0
- Change 1
- Change 2
  - Change 2.1

## v1.0.0
- Change 0
`, '2.0.0...3.0.0')

    expect(changes).toMatchSnapshot()
})

it('Can get changes after the latest', () => {
    const changes = extractChanges('file.md', `# Changelog

## v2.0.0
- Change 1
- Change 2
  - Change 2.1

## v1.0.0
- Change 0
`, '2.0.0...')

    expect(changes).toMatchSnapshot()
})

it('Can get changes for vNext', () => {
    const changes = extractChanges('file.md', `# Changelog

## v2.0.0
- Change 1
- Change 2
  - Change 2.1

## v1.0.0
- Change 0
`, '[2.0.0]...')

    expect(changes).toMatchSnapshot()
})
