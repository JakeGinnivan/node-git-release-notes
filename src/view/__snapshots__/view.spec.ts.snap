// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Can get changes after a version 1`] = `
"- Change 1
- Change 2
  - Change 2.1

- Change 3
- Change 4
"
`;

exports[`Can get changes before a version 1`] = `
"- Change 0
- Change 1
- Change 2
  - Change 2.1
"
`;

exports[`Can get changes between versions 1`] = `
"- Change 1
- Change 2
  - Change 2.1

- Fix 1
"
`;

exports[`Can get changes for a single version 1`] = `
"- Change 1
- Change 2
  - Change 2.1
"
`;
